### PR TITLE
Kill transposes 2

### PR DESCRIPTION
**Pull request by amilsted**
This adds a couple more tweaks to tensordot2 and also, perhaps more controversially, adds an `output_edge_ordering` argument to `contract_between()`. 

The latter allows `contract_between()` to flip the argument ordering in the call to `tensordot()` in case this would minimize transposition of the output tensor. For this to be possible, it must know the desired output ordering. This is used in `ncon()`.

Probably I should add some explicit tests for this is `tensornetwork_test`. Am happy to do that.

### CODE REVIEW

**Comment by Thenerdstation** [comment]
> perhaps more controversially, adds an output_edge_ordering argument to contract_between().

Doesn't seem controversial at all to me. 
>Probably I should add some explicit tests for this is tensornetwork_test. Am happy to do that.

Yes, always add tests. :)

**Comment by amilsted** [comment]
> Doesn't seem controversial at all to me.

Good :) It irked me a little to add a second way of achieving an output ordering, but I guess it's not like we're adhering to "TOOWTDI"* elsewhere.

Tests are in.

*just found out there's an acronym for this...

**Comment by Thenerdstation** [comment]
Well, the reason it doesn't break TOOWTDI is because you only do it if you have to. It's a necessary optimization.

File: tensornetwork/backends/tensorflow/tensordot2.py [hunk 1]
```diff
@@ -72,10 +72,30 @@ def _tensordot_should_flip(contraction_axes, free_axes):
       should_flip: `True` if `contraction_axes` should be moved to the left,
         `False` if they should be moved to the right.
     """
+    # NOTE: This will fail if the arguments contain any Tensors.
     if contraction_axes and free_axes:
       return bool(np.mean(contraction_axes) < np.mean(free_axes))
     return False
 
+  def _tranpose_if_necessary(tensor, perm):
```
Comment (Thenerdstation) [inline tensornetwork/backends/tensorflow/tensordot2.py]: Please add typing.

File: tensornetwork/backends/tensorflow/tensordot2.py [hunk 2]
```diff
@@ -72,10 +72,30 @@ def _tensordot_should_flip(contraction_axes, free_axes):
...
+    """Like transpose(), but avoids creating a new tensor if possible.
+    Although the graph optimizer should kill trivial transposes, it is best not
+    to add them in the first place!
+    """
+    # If perm is tensor-valued, test will fail and we fall back to tranposing.
+    if perm == list(range(len(perm))):
+      return tensor
+    return tf.transpose(tensor, perm)
+
+  def _reshape_if_necessary(tensor, new_shape):
```
Comment (Thenerdstation) [inline tensornetwork/backends/tensorflow/tensordot2.py]: Typing.

File: tensornetwork/backends/tensorflow/tensordot2.py [hunk 3]
```diff
@@ -99,37 +120,34 @@ def _tensordot_reshape(a, axes, is_right_term=False):
     """
     if a.get_shape().is_fully_defined() and isinstance(axes, (list, tuple)):
       shape_a = a.get_shape().as_list()
+      # NOTE: This will fail if axes contains any tensors
```
Comment (Thenerdstation) [inline tensornetwork/backends/tensorflow/tensordot2.py]: I don't think we should ever have axes be tensors, but if you add typing we can make sure that never happens ;)

File: tensornetwork/backends/tensorflow/tensordot2.py [hunk 4]
```diff
@@ -180,13 +203,15 @@ def _tensordot_axes(a, axes):
           isinstance(b_axes, tf.compat.integral_types):
         a_axes = [a_axes]
         b_axes = [b_axes]
+      # NOTE: This fails if either a_axes and b_axes are Tensors.
```
Comment (Thenerdstation) [inline tensornetwork/backends/tensorflow/tensordot2.py]: Axes will likely never be Tensors, so go ahead and remove this comment.

File: tensornetwork/network.py [hunk 5]
```diff
@@ -713,13 +717,40 @@ def contract_between(
...
 
+    if output_edge_order:
+      # Determine heuristically if output transposition can be minimized by 
+      # flipping the arguments to tensordot.
+      node1_output_axes = []
+      node2_output_axes = []
+      for (i, edge) in enumerate(output_edge_order):
+        if edge in shared_edges:
+          raise ValueError(
+            "Edge '{}' in output_edge_order is shared by the nodes to be "
```
Comment (Thenerdstation) [inline tensornetwork/network.py]: Can you add a test for this case?

File: tensornetwork/network.py [hunk 6]
```diff
@@ -713,13 +717,40 @@ def contract_between(
...
+            "contracted: '{}' and '{}'.".format(edge, node1, node2)
+          )
+        edge_nodes = set(edge.get_nodes())
+        if node1 in edge_nodes:
+          node1_output_axes.append(i)
+        elif node2 in edge_nodes:
+          node2_output_axes.append(i)
+        else:
+          raise ValueError(
+            "Edge '{}' in output_edge_order is not connected to node '{}' or "
```
Comment (Thenerdstation) [inline tensornetwork/network.py]: Add test for this case too. Generally, strive to have 100% test converge when ever possible. It makes refactoring and fixing bugs waaaaayyyyy easier in the long run.

**Comment by amilsted** [comment]
So, I've been trying to keep the tensordot2 code close to the TF tensordot code so that it can be pushed upstream at some point. I could add typing, but the TF code does not use it. This also affects the assumptions I can make about input arguments. For example, tensordot supports `axes` being a single Tensor, but TensorNetwork never uses this.

I can certainly still add typing to tensordot2 if you deem it necessary.

File: tensornetwork/network.py (hunk 5 above)
Comment (amilsted) [reply tensornetwork/network.py]: This is covered by the test I added.

File: tensornetwork/network.py (hunk 6 above)
Comment (amilsted) [reply tensornetwork/network.py]: This is also covered by the test I added.

File: tensornetwork/backends/tensorflow/tensordot2.py (hunk 1 above)
Comment (amilsted) [reply tensornetwork/backends/tensorflow/tensordot2.py]: See main thread.

**Comment by Thenerdstation** [comment]
Go ahead and add the typing. Removing it when you make an upstream PR should be pretty easy. 

If you want to include `tf.Tensor` as one of the types to `axes`, I would be ok with that.

**Comment by Thenerdstation** [comment]
Oh no a flaky test.

Why did I use np.random. let me fix that

**Pull request merged**